File: CmnClsLib_Func/TextOperation.vb

Imports System
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Windows.Forms


'使用Dictionary类需要的命名空间
Imports System.Collections.Generic
'使用Linq查询类需要的命名空间。
Imports System.Linq


Namespace CmnClsLib_Func
    Public Class TextOperation


        ''' <summary>
        ''' 根据给定的编码，确定字符串的byte长度。如“Shift-JIS“
        ''' </summary>
        ''' <param name="str">要确定长度的字符串</param>
        ''' <param name="charSetName">字符集的准确全称</param>
        ''' <returns></returns>
        Public Function GetStringLengthByByte(str As String, charSetName As String) As Integer
            Dim byteLength As Integer
            Select Case charSetName
                Case "UTF8"
                    byteLength = System.Text.Encoding.UTF8.GetBytes(str).Length
                    Exit Select
                Case "Unicode"
                    byteLength = System.Text.Encoding.Unicode.GetBytes(str).Length
                    Exit Select
                Case "ASCII"
                    byteLength = System.Text.Encoding.ASCII.GetBytes(str).Length
                    Exit Select
                Case "UTF32"
                    byteLength = System.Text.Encoding.UTF32.GetBytes(str).Length
                    Exit Select
                Case "BigEndianUnicode"
                    byteLength = System.Text.Encoding.BigEndianUnicode.GetBytes(str).Length
                    Exit Select
                Case "UTF7"
                    byteLength = System.Text.Encoding.UTF7.GetBytes(str).Length
                    Exit Select
                Case Else
                    byteLength = System.Text.Encoding.GetEncoding(charSetName).GetBytes(str).Length
                    Exit Select
            End Select
            Return byteLength
        End Function


        ''' <summary>
        ''' 控件输入内容合法性检测
        ''' </summary>
        ''' <param name="objects">Object后需要加[]，表示是Object类型的一维数组，params后必须是一维数组，使用Control类需要System.Web.UI命名空间</param>
        ''' <returns></returns>
        Public Function MustInputCheckForObject(objects As Object()) As Boolean
            Dim isCheckPass As Boolean = True
            For Each obj In objects
                Select Case (obj.GetType()).Name
                    Case "TextBox"
                        Dim tbx As Windows.Forms.TextBox = CType(obj, Windows.Forms.TextBox)
                        If String.IsNullOrEmpty(tbx.Text) = True Then
                            isCheckPass = False
                            Exit Function
                        End If
                    Case "ComboBox"
                        Dim cmbBox As Windows.Forms.ComboBox = CType(obj, Windows.Forms.ComboBox)
                        If String.IsNullOrEmpty(cmbBox.SelectedValue.ToString()) = True Then
                            isCheckPass = False
                            Exit Function
                        End If
                    Case "Label"
                        Dim lbl As Windows.Forms.Label = CType(obj, Windows.Forms.Label)
                        If String.IsNullOrEmpty(lbl.Text) = True Then
                            isCheckPass = False
                            Exit Function
                        End If
                End Select
            Next
            Return isCheckPass
        End Function

        ''' <summary>
        ''' 将合法的Code39字符串增加一个校验位
        ''' </summary>
        ''' <param name="strNeedCheckDigit"></param>
        ''' <returns></returns>
        Public Function Code39Mod43Check(strNeedCheckDigit As String) As String

            Dim Code39Mod43Dic As Dictionary(Of String, Integer) = New Dictionary(Of String, Integer)()
            Code39Mod43Dic.Add("0", 0)
            Code39Mod43Dic.Add("1", 1)
            Code39Mod43Dic.Add("2", 2)
            Code39Mod43Dic.Add("3", 3)
            Code39Mod43Dic.Add("4", 4)
            Code39Mod43Dic.Add("5", 5)
            Code39Mod43Dic.Add("6", 6)
            Code39Mod43Dic.Add("7", 7)
            Code39Mod43Dic.Add("8", 8)
            Code39Mod43Dic.Add("9", 9)
            Code39Mod43Dic.Add("A", 10)
            Code39Mod43Dic.Add("B", 11)
            Code39Mod43Dic.Add("C", 12)
            Code39Mod43Dic.Add("D", 13)
            Code39Mod43Dic.Add("E", 14)
            Code39Mod43Dic.Add("F", 15)
            Code39Mod43Dic.Add("G", 16)
            Code39Mod43Dic.Add("H", 17)
            Code39Mod43Dic.Add("I", 18)
            Code39Mod43Dic.Add("J", 19)
            Code39Mod43Dic.Add("K", 20)
            Code39Mod43Dic.Add("L", 21)
            Code39Mod43Dic.Add("M", 22)
            Code39Mod43Dic.Add("N", 23)
            Code39Mod43Dic.Add("O", 24)
            Code39Mod43Dic.Add("P", 25)
            Code39Mod43Dic.Add("Q", 26)
            Code39Mod43Dic.Add("R", 27)
            Code39Mod43Dic.Add("S", 28)
            Code39Mod43Dic.Add("T", 29)
            Code39Mod43Dic.Add("U", 30)
            Code39Mod43Dic.Add("V", 31)
            Code39Mod43Dic.Add("W", 32)
            Code39Mod43Dic.Add("X", 33)
            Code39Mod43Dic.Add("Y", 34)
            Code39Mod43Dic.Add("Z", 35)
            Code39Mod43Dic.Add("-", 36)
            Code39Mod43Dic.Add(".", 37)
            Code39Mod43Dic.Add(" ", 38)
            Code39Mod43Dic.Add("$", 39)
            Code39Mod43Dic.Add("/", 40)
            Code39Mod43Dic.Add("+", 41)
            Code39Mod43Dic.Add("%", 42)

            strNeedCheckDigit = strNeedCheckDigit.ToUpper()        '将字符串改为大写
            ' 字符串中每个字符映射之和
            Dim mappingSum As Integer = 0
            For i As Integer = 0 To strNeedCheckDigit.Length
                mappingSum = mappingSum + Code39Mod43Dic(strNeedCheckDigit.Substring(i, 1))
            Next
            Dim modOfSum As Integer = mappingSum Mod 43                                 '将映射之和按照43取余
            Dim checkDigit As String = String.Empty


            '''/取得字典中Value对应的Key方法1，Foreach循环
            'foreach KeyValuePair < String, Int() > kvp In Code39Mod43Dic)

            'If kvp.Value = modOfSum Then

            '    checkDigit = kvp.Key.ToString()
            'End If
            'Next

            ''/取得字典中Value对应的Key方法2，Foreach循环
            For Each key In Code39Mod43Dic.Keys
                If Code39Mod43Dic(key) = modOfSum Then
                    checkDigit = key.ToString()
                End If
            Next


            '''/取得字典中Value对应的Key方法3，Linq取得所有Keys
            'var Keys = Code39Mod43Dic.Where(q >= q.Value == modOfSum).Select(q >= q.Key)         '取得余数对应的字符，作为校验位。


            '''/取得字典中Value对应的Key方法4，Linq取得所有Keys
            'Dim keyList As List(Of String) = (From q In Code39Mod43Dic
            '                                  Where q.Value = modOfSum
            '                                  Select q.Key).ToList(Of String)


            ''取得字典中Value对应的Key方法5，Linq取得第一个Key
            'checkDigit = Code39Mod43Dic.FirstOrDefault(q >= q.Value == modOfSum).Key.ToString()


            Dim checkedStr As String = "*" + strNeedCheckDigit + checkDigit.ToString() + "*"

                Return checkedStr
        End Function



        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="code39Char"></param>
        ''' <returns></returns>
        Public Function Code39Map(code39Char As String) As Integer
            Select Case code39Char
                Case "0" : Return 0
                Case "1" : Return 1
                Case "2" : Return 2
                Case "3" : Return 3
                Case "4" : Return 4
                Case "5" : Return 5
                Case "6" : Return 6
                Case "7" : Return 7
                Case "8" : Return 8
                Case "9" : Return 9
                Case "A" : Return 10
                Case "B" : Return 11
                Case "C" : Return 12
                Case "D" : Return 13
                Case "E" : Return 14
                Case "F" : Return 15
                Case "G" : Return 16
                Case "H" : Return 17
                Case "I" : Return 18
                Case "J" : Return 19
                Case "K" : Return 20
                Case "L" : Return 21
                Case "M" : Return 22
                Case "N" : Return 23
                Case "O" : Return 24
                Case "P" : Return 25
                Case "Q" : Return 26
                Case "R" : Return 27
                Case "S" : Return 28
                Case "T" : Return 29
                Case "U" : Return 30
                Case "V" : Return 31
                Case "W" : Return 32
                Case "X" : Return 33
                Case "Y" : Return 34
                Case "Z" : Return 35
                Case "-" : Return 36
                Case "." : Return 37
                Case " " : Return 38
                Case "$" : Return 39
                Case "/" : Return 40
                Case "+" : Return 41
                Case "%" : Return 42
                Case Else : Return 43
            End Select
        End Function

        Public Function Code39Map(code39Int As Integer) As String
            Select Case code39Int
                Case 0 : Return "0"
                Case 1 : Return "1"
                Case 2 : Return "2"
                Case 3 : Return "3"
                Case 4 : Return "4"
                Case 5 : Return "5"
                Case 6 : Return "6"
                Case 7 : Return "7"
                Case 8 : Return "8"
                Case 9 : Return "9"
                Case 10 : Return "A"
                Case 11 : Return "B"
                Case 12 : Return "C"
                Case 13 : Return "D"
                Case 14 : Return "E"
                Case 15 : Return "F"
                Case 16 : Return "G"
                Case 17 : Return "H"
                Case 18 : Return "I"
                Case 19 : Return "J"
                Case 20 : Return "K"
                Case 21 : Return "L"
                Case 22 : Return "M"
                Case 23 : Return "N"
                Case 24 : Return "O"
                Case 25 : Return "P"
                Case 26 : Return "Q"
                Case 27 : Return "R"
                Case 28 : Return "S"
                Case 29 : Return "T"
                Case 30 : Return "U"
                Case 31 : Return "V"
                Case 32 : Return "W"
                Case 33 : Return "X"
                Case 34 : Return "Y"
                Case 35 : Return "Z"
                Case 36 : Return "-"
                Case 37 : Return "."
                Case 38 : Return " "
                Case 39 : Return "$"
                Case 40 : Return "/"
                Case 41 : Return "+"
                Case 42 : Return "%"
                Case Else : Return "ERROR"
            End Select
        End Function

    End Class

End Namespace

File: CmnClsLib_Func/TextOperation.cs
using System;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Collections.Generic;   // 使用Dictionary类需要的命名空间。
using System.Linq;                           // 使用Linq查询类需要的命名空间。




namespace LXL.CommonClassLib
{
    public class TextOperation
    {
        /// <summary>
        /// 根据给定的编码，确定字符串的byte长度。如“Shift-JIS“
        /// </summary>
        /// <param name="str">要确定长度的字符串</param>
        /// <param name="charSetName">字符集的准确全称</param>
        /// <returns></returns>
        public static int GetStringLengthByByte(string str, string charSetName)
        {
            int byteLength;
            switch (charSetName)
            {
                case "UTF8":
                    byteLength = System.Text.Encoding.UTF8.GetBytes(str).Length;
                    break;
                case "Unicode":
                    byteLength = System.Text.Encoding.Unicode.GetBytes(str).Length;
                    break;
                case "ASCII":
                    byteLength = System.Text.Encoding.ASCII.GetBytes(str).Length;
                    break;
                case "UTF32":
                    byteLength = System.Text.Encoding.UTF32.GetBytes(str).Length;
                    break;
                case "BigEndianUnicode":
                    byteLength = System.Text.Encoding.BigEndianUnicode.GetBytes(str).Length;
                    break;
                case "UTF7":
                    byteLength = System.Text.Encoding.UTF7.GetBytes(str).Length;
                    break;
                default:
                    byteLength = System.Text.Encoding.GetEncoding(charSetName).GetBytes(str).Length;
                    break;
            }
            return byteLength;
        }

        /// <summary>
        /// 控件输入内容合法性检测
        /// </summary>
        /// <param name="objects">Object后需要加[]，表示是Object类型的一维数组，params后必须是一维数组，使用Control类需要System.Web.UI;命名空间</param>
        /// <returns></returns>
        public static bool MustInputCheckForObject(params Object[] objects)
        {
            bool isCheckPass = true;

            foreach (Object obj in objects)
            {
                switch ((obj.GetType()).Name)
                {
                    case "TextBox":

                        TextBox tbx = (TextBox)obj;
                        if (String.IsNullOrEmpty(tbx.Text) == true)
                            isCheckPass = false;
                        tbx.Text = "tttttttttttttttttttttt";

                        break;
                    case "DropDownList":
                        DropDownList ddl = (DropDownList)obj;
                        if (String.IsNullOrEmpty(ddl.SelectedValue) == true)
                            isCheckPass = false;
                        ddl.Items.Add("aaa");
                        break;
                    case "Label":
                        Label lbl = (Label)obj;
                        if (String.IsNullOrEmpty(lbl.Text) == true)
                            isCheckPass = false;
                        lbl.Text = "llllllllllllllllllll";
                        break;
                }

            }
            return isCheckPass;


        }

        /// <summary>
        /// 将合法的Code39字符串增加一个校验位
        /// </summary>
        /// <param name="strNeedCheckDigit"></param>
        /// <returns></returns>
        public static string Code39Mod43Check(string strNeedCheckDigit)
        {
            Dictionary<string, int> Code39Mod43Dic = new Dictionary<string, int>();
            Code39Mod43Dic.Add("0", 0);
            Code39Mod43Dic.Add("1", 1);
            Code39Mod43Dic.Add("2", 2);
            Code39Mod43Dic.Add("3", 3);
            Code39Mod43Dic.Add("4", 4);
            Code39Mod43Dic.Add("5", 5);
            Code39Mod43Dic.Add("6", 6);
            Code39Mod43Dic.Add("7", 7);
            Code39Mod43Dic.Add("8", 8);
            Code39Mod43Dic.Add("9", 9);
            Code39Mod43Dic.Add("A", 10);
            Code39Mod43Dic.Add("B", 11);
            Code39Mod43Dic.Add("C", 12);
            Code39Mod43Dic.Add("D", 13);
            Code39Mod43Dic.Add("E", 14);
            Code39Mod43Dic.Add("F", 15);
            Code39Mod43Dic.Add("G", 16);
            Code39Mod43Dic.Add("H", 17);
            Code39Mod43Dic.Add("I", 18);
            Code39Mod43Dic.Add("J", 19);
            Code39Mod43Dic.Add("K", 20);
            Code39Mod43Dic.Add("L", 21);
            Code39Mod43Dic.Add("M", 22);
            Code39Mod43Dic.Add("N", 23);
            Code39Mod43Dic.Add("O", 24);
            Code39Mod43Dic.Add("P", 25);
            Code39Mod43Dic.Add("Q", 26);
            Code39Mod43Dic.Add("R", 27);
            Code39Mod43Dic.Add("S", 28);
            Code39Mod43Dic.Add("T", 29);
            Code39Mod43Dic.Add("U", 30);
            Code39Mod43Dic.Add("V", 31);
            Code39Mod43Dic.Add("W", 32);
            Code39Mod43Dic.Add("X", 33);
            Code39Mod43Dic.Add("Y", 34);
            Code39Mod43Dic.Add("Z", 35);
            Code39Mod43Dic.Add("-", 36);
            Code39Mod43Dic.Add(".", 37);
            Code39Mod43Dic.Add(" ", 38);
            Code39Mod43Dic.Add("$", 39);
            Code39Mod43Dic.Add("/", 40);
            Code39Mod43Dic.Add("+", 41);
            Code39Mod43Dic.Add("%", 42);



            strNeedCheckDigit = strNeedCheckDigit.ToUpper();        //将字符串改为大写

            int mappingSum = 0;                                                         //字符串中每个字符映射之和

            for (int i = 0; i < strNeedCheckDigit.Length; i++)
            {
                mappingSum = mappingSum + Code39Mod43Dic[strNeedCheckDigit.Substring(i, 1)];
            }

            int modOfSum = mappingSum % 43;                                 //将映射之和按照43取余
            string checkDigit = string.Empty;



            ////取得字典中Value对应的Key：方法1，Foreach循环
            //foreach (KeyValuePair<string, int> kvp in Code39Mod43Dic)
            //{
            //    if (kvp.Value == modOfSum)
            //    {
            //        checkDigit = kvp.Key.ToString();
            //    }
            //}

            ////取得字典中Value对应的Key：方法2，Foreach循环
            //foreach (string key in Code39Mod43Dic.Keys)
            //{
            //    if (Code39Mod43Dic[key] == modOfSum)
            //    {
            //        checkDigit = key.ToString();
            //    }
            //}


            ////取得字典中Value对应的Key：方法3，Linq取得所有Keys
            //var keys = Code39Mod43Dic.Where(q => q.Value == modOfSum).Select(q => q.Key);         //取得余数对应的字符，作为校验位。


            ////取得字典中Value对应的Key：方法4，Linq取得所有Keys
            //List<string> keyList = (from q in Code39Mod43Dic
            //                        where q.Value == modOfSum
            //                        select q.Key).ToList<string>();


            //取得字典中Value对应的Key：方法5，Linq取得第一个Key
            checkDigit = Code39Mod43Dic.FirstOrDefault(q => q.Value == modOfSum).Key.ToString();



            string checkedStr = "*" + strNeedCheckDigit + checkDigit.ToString() + "*";

            return checkedStr;
        }




        public static int Code39Map(string code39Char)
        {
            switch (code39Char)
            {
                case "0": return 0;
                case "1": return 1;
                case "2": return 2;
                case "3": return 3;
                case "4": return 4;
                case "5": return 5;
                case "6": return 6;
                case "7": return 7;
                case "8": return 8;
                case "9": return 9;
                case "A": return 10;
                case "B": return 11;
                case "C": return 12;
                case "D": return 13;
                case "E": return 14;
                case "F": return 15;
                case "G": return 16;
                case "H": return 17;
                case "I": return 18;
                case "J": return 19;
                case "K": return 20;
                case "L": return 21;
                case "M": return 22;
                case "N": return 23;
                case "O": return 24;
                case "P": return 25;
                case "Q": return 26;
                case "R": return 27;
                case "S": return 28;
                case "T": return 29;
                case "U": return 30;
                case "V": return 31;
                case "W": return 32;
                case "X": return 33;
                case "Y": return 34;
                case "Z": return 35;
                case "-": return 36;
                case ".": return 37;
                case " ": return 38;
                case "$": return 39;
                case "/": return 40;
                case "+": return 41;
                case "%": return 42;
                default: return 43;
            }            
        }

        public static string Code39Map(int code39Int)
        {
            switch (code39Int)
            {
                case 0: return "0";
                case 1: return "1";
                case 2: return "2";
                case 3: return "3";
                case 4: return "4";
                case 5: return "5";
                case 6: return "6";
                case 7: return "7";
                case 8: return "8";
                case 9: return "9";
                case 10: return "A";
                case 11: return "B";
                case 12: return "C";
                case 13: return "D";
                case 14: return "E";
                case 15: return "F";
                case 16: return "G";
                case 17: return "H";
                case 18: return "I";
                case 19: return "J";
                case 20: return "K";
                case 21: return "L";
                case 22: return "M";
                case 23: return "N";
                case 24: return "O";
                case 25: return "P";
                case 26: return "Q";
                case 27: return "R";
                case 28: return "S";
                case 29: return "T";
                case 30: return "U";
                case 31: return "V";
                case 32: return "W";
                case 33: return "X";
                case 34: return "Y";
                case 35: return "Z";
                case 36: return "-";
                case 37: return ".";
                case 38: return " ";
                case 39: return "$";
                case 40: return "/";
                case 41: return "+";
                case 42: return "%";
                default: return "ERROR";
            }
        }

    }
}


